homeassistant/components/backblaze/backup.py

```diff
@@ -0,0 +1,678 @@
+            )
+
+            # --- Metadata File Upload ---
+            _LOGGER.info("Uploading metadata file: %s", prefixed_metadata_filename)
+            await self._hass.async_add_executor_job(
+                lambda: self._bucket.upload_bytes(
+                    metadata_content_bytes,
+                    prefixed_metadata_filename,
+                    content_type="application/json",
+                    file_info={"metadata_only": "true"},
+                )
+            )
+            _LOGGER.info(
+                "Metadata file upload finished for %s", prefixed_metadata_filename
+            )
+
+        except B2Error as err:
+            _LOGGER.error("Backblaze B2 API error during backup upload: %s", err)
+            await self._cleanup_failed_upload(prefixed_tar_filename)
+            raise BackupAgentError(
+                f"Failed to upload backup to Backblaze B2: {err}"
+            ) from err
+        except Exception as err:
+            _LOGGER.exception("An unexpected error occurred during backup upload")
+            raise BackupAgentError(
+                f"An unexpected error occurred during backup upload: {err}"
+            ) from err
+        else:
+            _LOGGER.info("Backup upload complete: %s", prefixed_tar_filename)
+            # Invalidate cache after successful upload
+            self._invalidate_caches(
+                backup.backup_id, prefixed_tar_filename, prefixed_metadata_filename
+            )
+
+    async def _upload_backup_file(
+        self,
+        filename: str,
+        open_stream: Callable[[], Coroutine[Any, Any, AsyncIterator[bytes]]],
+        file_info: dict[str, Any],
+    ) -> None:
+        """Upload backup file to B2 using streaming."""
+        _LOGGER.debug("Starting streaming upload for %s", filename)
+
+        stream = await open_stream()
+        reader = AsyncIteratorReader(self._hass, stream)
+
+        try:
+            _LOGGER.info("Uploading backup file %s with streaming", filename)
+
+            content_type, _ = mimetypes.guess_type(filename)
+            file_version = await self._hass.async_add_executor_job(
+                lambda: self._bucket.upload_unbound_stream(
+                    reader,
+                    filename,
+                    content_type=content_type or "application/x-tar",
+                    file_info=file_info,
+                )
+            )
+            _LOGGER.info(
+                "Successfully uploaded %s (ID: %s)", filename, file_version.id_
+            )
+
+        except B2Error:
+            _LOGGER.exception("B2 connection error during upload for %s", filename)
+            raise
+        except Exception:
+            _LOGGER.exception("An error occurred during upload for %s:", filename)
+            raise
+        finally:
+            # Close the reader to clean up resources
+            reader.close()
+            await self._close_stream(stream)
```

Comment:
I don't think we need to call `aclose` on the object returned by the `open_stream` function. Why did you add it? Did you see resource leakage if you don't call it?